x/btcstaking/keeper/params.go

```diff
@@ -2,34 +2,134 @@ package keeper
+	binary.BigEndian.PutUint32(buf[:], v)
+	return buf[:]
+}
+func uint32FromBytes(b []byte) (uint32, error) {
+	if len(b) != 4 {
+		return 0, fmt.Errorf("invalid uint32 bytes length: %d", len(b))
+	}
+
+	return binary.BigEndian.Uint32(b), nil
+}
+
+func mustUint32FromBytes(b []byte) uint32 {
+	v, err := uint32FromBytes(b)
+	if err != nil {
+		panic(err)
+	}
+
+	return v
+}
+
+func (k Keeper) paramsStore(ctx context.Context) prefix.Store {
+	storeAdapter := runtime.KVStoreAdapter(k.storeService.OpenKVStore(ctx))
+	return prefix.NewStore(storeAdapter, types.ParamsKey)
+}
+
+func (k Keeper) nextParamsVersion(ctx context.Context) uint32 {
+	paramsStore := k.paramsStore(ctx)
+	it := paramsStore.ReverseIterator(nil, nil)
+	defer it.Close()
+
+	if !it.Valid() {
+		return 0
+	}
+
+	return mustUint32FromBytes(it.Key()) + 1
+}
+
+func (k Keeper) getLastParams(ctx context.Context) *types.StoredParams {
+	paramsStore := k.paramsStore(ctx)
+	it := paramsStore.ReverseIterator(nil, nil)
+	defer it.Close()
+
+	if !it.Valid() {
+		return nil
+	}
+	var sp types.StoredParams
+	k.cdc.MustUnmarshal(it.Value(), &sp)
+	return &sp
+}
+
 // SetParams sets the x/btcstaking module parameters.
 func (k Keeper) SetParams(ctx context.Context, p types.Params) error {
 	if err := p.Validate(); err != nil {
 		return err
 	}
-	store := k.storeService.OpenKVStore(ctx)
-	bz := k.cdc.MustMarshal(&p)
-	return store.Set(types.ParamsKey, bz)
+
+	nextVersion := k.nextParamsVersion(ctx)
+	paramsStore := k.paramsStore(ctx)
+
+	sp := types.StoredParams{
+		Params:  p,
+		Version: nextVersion,
+	}
+
+	paramsStore.Set(uint32ToBytes(nextVersion), k.cdc.MustMarshal(&sp))
+	return nil
 }
 
-// GetParams returns the current x/btcstaking module parameters.
+// GetParams returns the latest x/btcstaking module parameters.
 func (k Keeper) GetParams(ctx context.Context) (p types.Params) {
-	store := k.storeService.OpenKVStore(ctx)
-	bz, err := store.Get(types.ParamsKey)
-	if err != nil {
-		panic(err)
-	}
-	if bz == nil {
+	lastParams := k.getLastParams(ctx)
+
+	if lastParams == nil {
 		return p
 	}
-	k.cdc.MustUnmarshal(bz, &p)
+
+	return lastParams.Params
+}
+
+func (k Keeper) GetAllParams(ctx context.Context) []*types.Params {
+	paramsStore := k.paramsStore(ctx)
+	it := paramsStore.Iterator(nil, nil)
+	defer it.Close()
+
+	var p []*types.Params
+	for ; it.Valid(); it.Next() {
+		var sp types.StoredParams
+		k.cdc.MustUnmarshal(it.Value(), &sp)
+		p = append(p, &sp.Params)
+	}
+
 	return p
 }
 
+func (k Keeper) GetParamsByVersion(ctx context.Context, v uint32) *types.Params {
+	paramsStore := k.paramsStore(ctx)
+	spBytes := paramsStore.Get(uint32ToBytes(v))
+	if len(spBytes) == 0 {
+		return nil
+	}
+
+	var sp types.StoredParams
+	k.cdc.MustUnmarshal(spBytes, &sp)
+	return &sp.Params
+}
+
+func (k Keeper) GetParamsWithVersion(ctx context.Context) (sp types.StoredParams) {
+	lastParams := k.getLastParams(ctx)
+
+	if lastParams == nil {
+		return sp
+	}
+
+	return *k.getLastParams(ctx)
+}
```

Comment:
```suggestion
func (k Keeper) GetParamsWithVersion(ctx context.Context) *types.StoredParams {
	return k.getLastParams(ctx)
}
```
Also how about returning a `Params` and a version rather than returning `StoredParams` directly? this way `StoredParams` is entirely internal for `params.go`

Reply:
tbh I dislike returning tuples from functions as returned values do not have the names, but maybe in this constrained context it makes sense 🤔 will see

Reply:
checked how it looks in code and I do not like too much, it makes this two values (params and version) a bit unconnected to each other.

Another way would be to introduce another struct `VersionedParams` which would be used only in operations (not db) but I am not sure it is worth it